### PR TITLE
[risk=no][RW-4458]create new workspace test

### PR DESCRIPTION
**Pull request by aweng98**

### CODE REVIEW

File: e2e/app/elements/button.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import {Page} from 'puppeteer';
-import * as elementHandler from '../../driver/element-handler';
+import * as elementHandler from '../../driver/elementHandle-util';
```
Comment (aweng98) [inline e2e/app/elements/button.ts]: file rename.

File: e2e/app/google-login.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
-import {ElementHandle, Page} from 'puppeteer';
-import {waitForNavigation} from '../services/page-wait';
+import { ElementHandle, Page } from 'puppeteer';
+import { waitForNavigation } from '../services/page-wait';
 
-const configs = require('../config/config');
+const configs = require('../resources/config');
```
Comment (aweng98) [inline e2e/app/google-login.ts]: file move.

File: e2e/app/workspace-element.ts [hunk 3]
```diff
@@ -23,32 +24,3 @@ exports.SELT_XPATH_QUESTION4 = `//*[starts-with(normalize-space(.), "4. What are
 // request a review Yes or No radiobutton
 exports.SELT_XPATH_YES = `//*[label[contains(normalize-space(.),"Would you like to request a review of your research purpose")]]/following-sibling::*/input[@type="radio"][following-sibling::label[1]/text()="Yes"]`;
 exports.SELT_XPATH_NO = `//*[label[contains(normalize-space(.),"Would you like to request a review of your research purpose")]]/following-sibling::*/input[@type="radio"][following-sibling::label[1]/text()="No"]`;
-
-export default class ProjectPurposeQuestion extends Widget {
```
Comment (aweng98) [inline e2e/app/workspace-element.ts]: renamed class name to `ComponentWebElement`

File: e2e/app/workspace-page.ts [hunk 4]
```diff
@@ -44,62 +52,62 @@ export default class Workspaces extends authenticatedpage {
     return await this.puppeteerPage.waitFor(selectors.dataSet, { visible: true });
   }
 
-  public element_question1_forProfitPurpose(): ProjectPurposeQuestion {
-    return new ProjectPurposeQuestion(this.puppeteerPage, 'For-Profit Purpose');
+  public question1ForProfit(): ComponentWebElement {
```
Comment (aweng98) [inline e2e/app/workspace-page.ts]: bunch of methods rename to make shorter

File: e2e/driver/AouElement.ts [hunk 5]
```diff
@@ -1,75 +1,115 @@
...
+   * If the attribute have property counterparts, use getProperty function.
    */
-  public async getAttributeAlternative(page: Page, element: ElementHandle, attribute: string) {
-    const handle = await page.evaluateHandle( (elem, attr) => {
-      return elem.getAttribute(attr);
-    }, element, attribute);
-    return await handle.jsonValue();
+  public async getAttr(attr: string): Promise<unknown> {
+    const atr = await this.eHandle.getAttribute(attr);
+    console.log(`getAttr(${attr}) function: ${atr}`);
```
Comment (als364) [inline e2e/driver/AouElement.ts]: Does this still debug statement still need to be there?

File: e2e/driver/AouElement.ts [hunk 6]
```diff
@@ -1,75 +1,115 @@
...
-
-   /**
-    * Cannot get attribute directly from a ElementHandle.
-    * If the attribute have property counterparts, use getProperty function.
-    */
-  public async getAttribute(page: Page) {
-    return await page.evaluate(link => link.getAttribute('href'), this.eHandle,);
+  public async getProp(property: string): Promise<unknown> {
+    const prop = await this.eHandle.getProperty(property);
+    console.log(`getProp(${property}) function: ${await prop.jsonValue()}`);
```
Comment (als364) [inline e2e/driver/AouElement.ts]: Does this still debug statement still need to be there?

File: e2e/driver/AouElement.ts [hunk 7]
```diff
@@ -1,75 +1,115 @@
...
+      await this.click();
     }
   }
 
   public async unCheck() {
-    if (this.isChecked()) {
-      this.click();
+    const cValue = await this.isChecked();
+    if (cValue) {
+      console.log('going to uncheck the checkbox');
```
Comment (als364) [inline e2e/driver/AouElement.ts]: Does this still debug statement still need to be there?

File: e2e/driver/AouElement.ts [hunk 8]
```diff
@@ -1,75 +1,115 @@
...
+  }
+
+  public async value(): Promise<string> {
+    const p = await this.getProp('value');
+    if (typeof p === 'undefined') { return p; } else { return ''; }
+  }
+
+  public async isChecked(): Promise<boolean> {
+    const checkedProp = await this.getProp('checked');
+    console.log('isChecked function returned: ' + checkedProp);
```
Comment (als364) [inline e2e/driver/AouElement.ts]: Does this still debug statement still need to be there?

File: e2e/driver/AouElement.ts (hunk 5 above)
Comment (aweng98) [reply e2e/driver/AouElement.ts]: I need to see logged info for now until common functions are working consistently. eventually console.log will be all removed.

File: e2e/app/mixin/authenticatedpage.ts [hunk 9]
```diff
@@ -43,17 +43,23 @@ export default abstract class AuthenticatedPage extends BasePage {
   protected async waitForSpinner() {
     // wait maximum 1 second for either spinner to show up
     const selectr1 = '.spinner, svg';
-    const found = await this.puppeteerPage.waitFor((selector) => {
+    const spinner = await this.puppeteerPage.waitFor((selector) => {
       return document.querySelectorAll(selector).length > 0
     }, {timeout: 1000}, selectr1);
+    const f = await spinner.jsonValue();
```
[the remark this answers is not in the document]
Comment (aweng98) [reply e2e/app/mixin/authenticatedpage.ts]: renamed to jValue.

File: e2e/app/mixin/authenticatedpage.ts [hunk 10]
```diff
@@ -43,17 +43,23 @@ export default abstract class AuthenticatedPage extends BasePage {
...
     const selectr2 = 'svg[style*="spin"], .spinner:empty';
-    if (found) {
+
+    if (f) {
       await this.puppeteerPage.waitFor((selector) => {
         return document.querySelectorAll(selector).length === 0
       }, {timeout: 60000}, selectr2);
     }
+    // final 1 second wait for page render to finish
+    if (f){
```
[the remark this answers is not in the document]
Comment (aweng98) [reply e2e/app/mixin/authenticatedpage.ts]: fixed

File: e2e/app/workspace-page.ts (hunk 4 above)
Comment (aweng98) [reply e2e/app/workspace-page.ts]: for grouping get-element functions. too many elements on page. naming convention probably will change in future.

**Comment by aweng98** [comment]
@als364 Thank you.

**Pull request merged**